pkg/apiserver/registry/core/shoot/strategy_test.go

```diff
@@ -65,7 +67,77 @@ var _ = Describe("Strategy", func() {
 		})
 	})
 
+	Describe("#PrepareForCreate", func() {
+		Context("cloudProfile field removal", func() {
+			var (
+				shoot                 *core.Shoot
+				cloudProfileReference *core.CloudProfileReference
+			)
+
+			BeforeEach(func() {
+				shoot = &core.Shoot{}
+				cloudProfileReference = &core.CloudProfileReference{
+					Kind: "foo",
+					Name: "bar",
+				}
+			})
+
+			It("should remove cloudProfile field if namespacedcloudprofile feature gate is disabled", func() {
+				DeferCleanup(test.WithFeatureGate(features.DefaultFeatureGate, features.UseNamespacedCloudProfile, false))
+
+				shoot.Spec.CloudProfile = cloudProfileReference
+				strategy.PrepareForCreate(context.TODO(), shoot)
+
+				Expect(shoot.Spec.CloudProfile).To(BeNil())
+			})
+
+			It("should not remove cloudProfile field if namespacedcloudprofile feature gate is enabled", func() {
+				DeferCleanup(test.WithFeatureGate(features.DefaultFeatureGate, features.UseNamespacedCloudProfile, true))
+
+				shoot.Spec.CloudProfile = cloudProfileReference
+				strategy.PrepareForCreate(context.TODO(), shoot)
+
+				Expect(shoot.Spec.CloudProfile).To(Equal(cloudProfileReference))
+			})
+		})
+	})
+
 	Describe("#PrepareForUpdate", func() {
+		Context("cloudProfile field removal", func() {
+			var (
+				oldShoot              *core.Shoot
+				newShoot              *core.Shoot
+				cloudProfileReference *core.CloudProfileReference
+			)
+
+			BeforeEach(func() {
+				oldShoot = &core.Shoot{}
+				newShoot = oldShoot.DeepCopy()
+				cloudProfileReference = &core.CloudProfileReference{
+					Kind: "foo",
+					Name: "bar",
+				}
+			})
+
+			It("should remove cloudProfile field if namespacedcloudprofile feature gate is disabled", func() {
+				DeferCleanup(test.WithFeatureGate(features.DefaultFeatureGate, features.UseNamespacedCloudProfile, false))
+
+				newShoot.Spec.CloudProfile = cloudProfileReference
+				strategy.PrepareForUpdate(context.TODO(), newShoot, oldShoot)
+
+				Expect(newShoot.Spec.CloudProfile).To(BeNil())
+			})
```

Comment:
Does it make sense to add a case when field is not nil in the old Shoot and hence we don't set it to nil?